### PR TITLE
Make IR ir.Del free until legalized.

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
This essentially prevents `ir.Del` from being present in the IR until it is legalised. It's fine for a pass to generate `ir.Del` nodes for whatever reason and then strip them before the pass completes. Strict enforcement is applied to passes inheriting from `numba.core.compiler_machinery.FunctionPass`.

**Comment by stuartarchibald** [comment]
CI is stuck because of a github outage, the close/open might un-stick it.

**Comment by sklam** [comment]
The following tests are failing on my osx machine
```pytb
ERROR: test_obj_func_match (numba.tests.test_ir_utils.TestIrUtils)
FAIL: test_dead_code_elimination (numba.tests.test_ir_utils.TestIrUtils)
```

File: numba/core/postproc.py [hunk 1]
```diff
@@ -209,3 +213,16 @@ def _patch_var_dels(self, internal_dead_map, escaping_dead_map):
             escape_dead_set = escaping_dead_map[offset]
             for var_name in sorted(escape_dead_set):
                 ir_block.prepend(ir.Del(var_name, loc=ir_block.body[0].loc))
+
+
+    def remove_dels(self):
+        """
+        Strips the IR of Del nodes
+        """
+        for block in self.func_ir.blocks.values():
```
Comment (sklam) [inline numba/core/postproc.py]: this can call `ir_utils.remove_dels` instead

File: numba/tests/test_ir.py [hunk 2]
```diff
@@ -431,6 +436,7 @@ def check_diffstr(string, pointing_at=[]):
                 b[idx], b[idx + 1] = b[idx + 1], b[idx]
                 break
 
+        assert change_set
```
Comment (sklam) [inline numba/tests/test_ir.py]: use `self.assertTrue` instead

File: numba/core/ir_utils.py [hunk 3]
```diff
@@ -2056,20 +2056,29 @@ def resolve_mod(mod):
         return None
 
 
+def enforce_no_dels(func_ir):
+    """
+    Enforce there being no ir.Del nodes in the IR.
+    """
+    strict = True
```
Comment (sklam) [inline numba/core/ir_utils.py]: remove this option or make it configurable

File: numba/core/ir_utils.py [hunk 4]
```diff
@@ -2056,20 +2056,29 @@ def resolve_mod(mod):
...
+    for blk in func_ir.blocks.values():
+        dels = [x for x in blk.find_insts(ir.Del)]
+        if dels:
+            msg = "Illegal IR, del found at: %s" % dels[0]
+            if strict:
+                raise CompilerError(msg, loc=dels[0].loc)
```
Comment (sklam) [inline numba/core/ir_utils.py]: Can we have a direct test for this function?

**Comment by stuartarchibald** [comment]
Bizarrely https://github.com/numba/numba/pull/5271/commits/7c02c07daa9335e37c6ee51f4a49a72c2170b35b passed, which is identical to this branch apart from a dead import statement.

File: numba/tests/test_ir.py (hunk 2 above)
Comment (stuartarchibald) [reply numba/tests/test_ir.py]: can do, this isn't actually part of the testing, it's ensuring that the state of the thing to test is actually valid.

File: numba/core/ir_utils.py (hunk 3 above)
Comment (stuartarchibald) [reply numba/core/ir_utils.py]: Opt to remove. The pass manager is enforcing.

**Comment by stuartarchibald** [comment]
> Bizarrely [7c02c07](https://github.com/numba/numba/commit/7c02c07daa9335e37c6ee51f4a49a72c2170b35b) passed, which is identical to this branch apart from a dead import statement.

Rerun is showing the failures you saw @sklam, I guess github issues yesterday lead to a false pass.

File: numba/core/ir_utils.py (hunk 4 above)
Comment (stuartarchibald) [reply numba/core/ir_utils.py]: done in e977137

**Comment by stuartarchibald** [comment]
> The following tests are failing on my osx machine
> 
> ```
> ERROR: test_obj_func_match (numba.tests.test_ir_utils.TestIrUtils)
> FAIL: test_dead_code_elimination (numba.tests.test_ir_utils.TestIrUtils)
> ```
> 
> which are the same failures as in azure

Hopefully now fixed. Tests were checking `ir.Del` counts.

**Comment by stuartarchibald** [comment]
@sklam thanks for the review, I think I've addressed everything.

**Pull request merged**